From 2e7d1278b18930e4f175c320e82a0f4db697a77e Mon Sep 17 00:00:00 2001
From: Luc Maisonobe <luc@apache.org>
Date: Tue, 2 Apr 2013 19:02:55 +0000
Subject: [PATCH] Fixed missing side effects of secondary equations.

When secondary equations have a side effect on main state in Ordinary
Differential Equations integration, it is now preoerly taken into
account.

JIRA: MATH-960

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/math/trunk@1463680 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                                       | 4 ++++
 .../org/apache/commons/math3/ode/ExpandableStatefulODE.java   | 3 ++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 6521f7f936..c12a2a6cc1 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -78,6 +78,10 @@ Users are encouraged to upgrade to this version as this release not
   2. A few methods in the FastMath class are in fact slower that their
   counterpart in either Math or StrictMath (cf. MATH-740 and MATH-901).
 ">
+      <action dev="luc" type="fix" issue="MATH-960" >
+        Fixed missing side effects of secondary equations on main state in
+        Ordinary Differential Equations integration.
+      </action>
       <action dev="luc" type="fix" issue="MATH-957" due-to="Evan Ward">
         Fixed inverse cumulative probability for uniform distribution.
       </action>
diff --git a/src/main/java/org/apache/commons/math3/ode/ExpandableStatefulODE.java b/src/main/java/org/apache/commons/math3/ode/ExpandableStatefulODE.java
index fe918537fc..842e58286d 100644
--- a/src/main/java/org/apache/commons/math3/ode/ExpandableStatefulODE.java
+++ b/src/main/java/org/apache/commons/math3/ode/ExpandableStatefulODE.java
@@ -117,7 +117,6 @@ public void computeDerivatives(final double t, final double[] y, final double[]
         // compute derivatives of the primary equations
         primaryMapper.extractEquationData(y, primaryState);
         primary.computeDerivatives(t, primaryState, primaryStateDot);
-        primaryMapper.insertEquationData(primaryStateDot, yDot);
 
         // Add contribution for secondary equations
         for (final SecondaryComponent component : components) {
@@ -127,6 +126,8 @@ public void computeDerivatives(final double t, final double[] y, final double[]
             component.mapper.insertEquationData(component.stateDot, yDot);
         }
 
+        primaryMapper.insertEquationData(primaryStateDot, yDot);
+
     }
 
     /** Add a set of secondary equations to be integrated along with the primary set.
